make footnotes from table bodies hyperlinked in PDF output

jfbu opened the pull request:
Subject: improving footnotes in tables

### Feature or Bugfix
- Feature

### Purpose
- currently the LaTeX writer uses `\footnotemark/\footnotetext` mark-up for all footnotes inside tables. (*) But in such cases ``hyperref``  creates no hyperlinks. This PR arranges for using ``\footnote`` mark-up when possible in order for the footnote marker to be hyperlinked.

(*) in fact since Sphinx 1.5 (b8e0e93), ``footnote`` and ``footnotetext`` environments are used in place of ``\footnote`` and ``\footnotetext`` macros and they allow verbatim content (code-blocks). This uses the LaTeX package ``footnotehyper-sphinx`` which is shipped with Sphinx and loads old package ``footnote.sty`` which it patches.


### Detail

For footnotes in table captions it is still needed that `mark/text` syntax be used; and, as longtable headers and footers are typeset more than once if the table extends over multiple pages, the `mark/text` must still be used there (else the footnote text would be inserted multiple times at bottom of last page.) Thus this PR arranges for the "restriction" on footnotes to be raised at entry of table and lifted at entrance of the table body. The accumulated footnote texts from caption and header are delivered at beginning of first cell of body, for the corresponding texts to appear prior to the ones from the table cells.

All table environments are wrapped into a "savenotes" environment. Earlier, Sphinx patched the ``tabular``, ``tabulary``, and the ``treeparttable``  environments via the ``\makesavenoteenv`` macro, but it is clearer, avoids some redundancy and facilitates LaTeX debugging when the use of "savenotes" is made explicit. Thanks to the table templates (#3389) this can be done easily.

``longtable`` by itself allows ``\footnote`` in table cells but not the ``\footnote[N]`` form (or rather since 1.5 ``\begin{footnote}[N]``) which Sphinx uses as it numerates itself the footnote. This is a known bug of ``longtable`` (which is not corrected by ``hyperref``'s patch of ``longtable``.) The ``\sphinxlongtablepatch`` macro fixes this (but only when ``longtable`` environment is wrapped in ``savenotes`` environment.)

**edit**: *the first version of ``\sphinxlongtablepatch`` worked by luck, because it was misformed. The new version at fca08d1 although simplistic is the correct one. The commit also makes a patch to make all forms of footnote syntax work well with hyperref inside ``savenotes`` and not only the syntax used by Sphinx latex writer.*

Currently, Sphinx ``footnotehyper-sphinx`` still loads ``footnote`` package but in future all macros will be defined directly rather than loading the old ``footnote`` package and patch all its macros to fix bugs and ensure ``hyperref`` compatibility.

**edit**:  *the LaTeX package ``footnotehyper`` has a new ``0.99`` version. In future ``footnotehyper-sphinx`` will be udpated to simply copy it. The ``0.99`` version does not anymore load 20 years old LaTeX package ``footnote``  with all the needed patches and fixes, and it defines all macros from scratch. This makes understanding code much easier.*

The use of "savenotes" environment means that the footnotes are delivered in PDF output after the table. This is already the case hence there is no change. For the case of ``longtable`` it would require much more LaTeX hacking to manage to get the hyperlinked footnotes delivered pagewise, if the table extends over multiple pages.

### Relates
- #2616, #3389

jfbu committed:
(latex) hyperlinked footnotes from table bodies

Memo: footnotehyper-sphinx LaTeX package incorporates here some fixes
from upstream footnotehyper 0.99. At Sphinx 1.6 il will for simplicity
copy all of upstream footnotehyper 0.99. Then old footnote.sty LaTeX
package is not a dependency of Sphinx anymore.

jfbu commented on sphinx/writers/latex.py:
```diff
@@ -1342,6 +1342,9 @@ def depart_thead(self, node):
         self.pushbody(self.table.body)  # Redirect body output until table is finished.
+        if self.footnote_restricted:
+            # releases footnotetexts from header in first non header cell
+            self.unrestrict_footnote(node.parent.parent)
```
at @tk0miya can you check if this is correct ? the aim is to deliver the footnotes which have been "restricted" since entrance in table. Due to the way ``unrestrict_ footnote`` works I must pass it correct node, and I hope ``node.parent.parent`` is the correct one in all cases. Those footnotes must be delivered first in table body to keep the ordering right.

tk0miya replied:
You're right. The table nodes are always structured as `table > tgroup > tbody`.
But I feel it is unbalance to start restriction on `visit_table` and to unrestrict on `visit_tbody`.
Basically, a symmetry is important to make readability.

To my understanding, the restriction is needed during processing a table caption.
So it would be nice like following:
```
def visit_title(self, node):
    ...
        elif isinstance(parent, nodes.table):  # table caption!
            # Redirect body output until title is finished.
            self.pushbody([])
            self.restrict_footnote(node)  # start footnote-restriction for table caption

def depart_title(self, node):
    ...
    if isinstance(node.parent, nodes.table):
        self.table.caption = self.popbody()  # store a caption text of the table

        self.pushbody([])   # add buffer for \footnotetext's temporarily
        self.unrestrict_footnote(node)
        self.table.caption_footnotes = self.popbody()  # store footnotes inside the caption

def depart_table(self, node):
    ...
    # output footnotes mark-ups on self.table.caption_footnote here
    # (using self.body.append() or templates)
    ...
```

What do you think?

jfbu replied:
Thanks for advice @tk0miya , can you check 3cbb555 ?

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -44,10 +44,7 @@
-\makesavenoteenv{tabular}
-\makesavenoteenv{threeparttable}
-% (longtable is hyperref compatible and needs no special treatment here.)
+\AtBeginDocument{\@ifpackageloaded{longtable}{\sphinxlongtablepatch}{}}%
```
the macro ``sphinxlongtablepatch`` is defined in  ``footnotehyper-sphinx``. Perhaps a comment should be added here.

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -44,10 +44,7 @@
 \RequirePackage{footnotehyper-sphinx}
-\makesavenoteenv{tabulary}
-\makesavenoteenv{tabular}
-\makesavenoteenv{threeparttable}
```
rather than using ``\makesavenoteenv`` which patches environments, it is safer to wrap these in to ``savenotes`` environment. For example with ``table`` environment (not used currently by Sphinx), I experienced that the ``\begin{table}[H]`` form redefines ``\endtable`` from inside ``\begin{table}`` causing a bug if ``\makesavenoteenv{table}`` has been used.

jfbu commented on sphinx/texinputs/footnotehyper-sphinx.sty:
```diff
@@ -87,7 +102,7 @@
     \let\@makefntext\@empty\let\@finalstrut\@gobble
     \let\rule\@gobbletwo
-    \if@savingnotes\expandafter\fn@fntext\else\expandafter\H@@footnotetext\fi
+    \if@savingnotes\expandafter\FNH@fn@fntext\else\expandafter\H@@footnotetext\fi
```
The `\H@@footnotetext` macro is the original `\@footnotetext` that `hyperref` keeps in addition to the `\@footnotetext` patched for hyperlinks. The `\FNH@fn@fntext` is the one which is used inside `savenotes` environment, it is the original `\fn@fntext` from footnote.sty package. But the`\fn@fntext` in earlier version referred to the redefined one by footnotehyper-sphinx, `\FNH@hyper@fntext` hence it was a bug here but escaped detection because Sphinx inserted all `footnotetext` environment after the tabular, tabulary, threeparttable, hence the footnote version of footnotehyper-sphinx was not executed, and the bug did not show. But if one starts locating the footnotetext inside the table, then the bug shows. This line fixes it.

jfbu commented on sphinx/texinputs/footnotehyper-sphinx.sty:
```diff
@@ -153,6 +169,21 @@
+\def\sphinxlongtablepatch {% only for longtable wrapped in "savenotes"
+    \def\LT@p@ftntext
+    {\ifx\FNH@endfntext@next\FNH@endfntext@nolink
+     \expandafter\FNH@fn@fntext\else\expandafter\FNH@hyper@fntext\fi }%
```
hmm.. the "False" branch is always taken due to misformed test. But this is correct for Sphinx anyhow. The test (from a dev version of upstream footnotehyper) was to avoid a `\@footnotetext` trying to do an hyperlink when none exists due to `\footnotemark[N]/\footnotetext[N]` case. But Sphinx uses environment form for `footnotetext` and never encounters `\@footnotetext` then. Hence the misformed test is not a bug for Sphinx. Nevertheless it is a bug which I will fix after getting feedback from upstream footnotehyper.

tk0miya commented:
I don't run this at my local yet. But it looks good to me.
Please go ahead. (see my comment above)

jfbu committed:
LaTeX writer: symmetrize footnote restrictions for tables

Memo 1: footnotes from captions are inserted by this commit via
table templates, so it is easy to put caption at foot of table
rather than at head, if wanted.

Memo 2: footnotes from table header need restriction due to
longtable only. But one can not tell in advance if table will be
rendered by longtable or not, hence one must restrict always.

jfbu commented:
I have rebased on master and incorporate @tk0miya advice about preserving symmetric usage of `restrict_footnotes(node)` in latex.py. If no objection I will merge in a few days. Thanks for reviewing !

tk0miya commented on tests/test_build_latex.py:
```diff
@@ -480,12 +480,11 @@ def test_footnote(app, status, warning):
-    assert ('\\end{threeparttable}\n\\par\n\\endgroup\n%\n'
-            '\\begin{footnotetext}[4]\\sphinxAtStartFootnote\n'
-            'footnotes in table caption\n%\n\\end{footnotetext}%\n'
+    assert ('\\begin{footnotetext}[4]\\sphinxAtStartFootnote\n'
```
We should confirm these `footnotetext`s are placed after the table environments.

jfbu replied:
thanks I will check but in fact now they are located inside table environment in first cell after header. This is needed so that footnotes from caption are not listed after footnotes from table cells. I will extend test to check correct positioning.

tk0miya commented on sphinx/writers/latex.py:
```diff
@@ -1333,15 +1340,27 @@ def depart_tgroup(self, node):
+        self.pushbody(self.table.body)
+        # insert footnotetexts from header at start of body (due to longtable)
+        # those from caption are handled by templates (to allow caption at foot)
+        self.body.extend(self.table.header_footnotetexts)
```
Is the test for `header_footnotetexts` needed?

jfbu replied:
Yes, a LaTeX test is needed in my opinion but if the case of footnotetexts from caption is treated it will probably be also test for header_footnotetexts. I don't think we need pure Python test on header_footnotetexts, only check that it is well inserted at start of table body. I will check.

tk0miya reviewed (approved):
LGTM with nits.

jfbu committed:
update root footnote test

jfbu committed:
update test_build_latex.py test of footnote in longtable caption

tk0miya commented:
LGTM!

jfbu commented:
Thank you for your reviewing, now merging !
